ENH Do not pass random_state to estimator in IterativeImputer

david-cortes opened the pull request:
Trying to fix issue: https://github.com/scikit-learn/scikit-learn/issues/15611

Lots of algorithms that rely on randomization offer an argument `random_state` through which it's possible to pass them RNG seeds that would ensure reproducible results. The `IterativeImputer` class will take some regressor/classifier and set their attribute `random_state` to a NumPy MT19937 object class if the object has such attribute, in such a way that this object is modified implicitly once the regressor/classifier to which it was assigned produces some random number.

This is fine for SciKit-Learn’s own classes, but other libraries which provide different regressors/classifiers which are SciKit-Learn-compatible might not be able to use such `RandomState` objects (e.g. if they generate random numbers in C++ with some method other than MT19937).

This PR addresses the issue by setting the `random_state` of the estimator to ~~a random integer, which is drawn from the imputer's `RandomState` object~~ the same value that was passed to the imputer, but does so only if it's not already set in the estimator. This way:
* It doesn't override the `random_state` which is passed to the constructor of the estimator used for imputation (if it was passed).
* Allows to use a wider array of SciKit-Learn-compatible objects from external libraries as estimators for `IterativeImputer`. Example:
```python
import numpy as np
from lightgbm import LGBMRegressor
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer

np.random.seed(123)
X = np.random.normal(size = (100, 5))
X[np.random.randint(100, size = 10),
  np.random.randint(5,   size = 10)] = np.nan

imputer = IterativeImputer(LGBMRegressor(), random_state = 456)
X_full = imputer.fit_transform(X)
```

rth commented on sklearn/impute/_iterative.py:
```diff
@@ -559,7 +559,8 @@ def fit_transform(self, X, y=None):
 
         if hasattr(self._estimator, 'random_state'):
-            self._estimator.random_state = self.random_state_
+            if self._estimator.random_state is None:
```
Shouldn't this be
```suggestion
            if self._estimator.random_state is None and self.random_state_ is not None:
```
?

david-cortes replied:
I don't think so, because `self.random_state_` (with underscore) is returned from function `check_random_state` through `self.random_state` (no underscore), so it cannot be missing (L541).

rth commented on sklearn/impute/_iterative.py:
```diff
@@ -559,7 +559,8 @@ def fit_transform(self, X, y=None):
         if hasattr(self._estimator, 'random_state'):
-            self._estimator.random_state = self.random_state_
+            if self._estimator.random_state is None:
+                self._estimator.random_state = self.random_state_.randint(np.iinfo(np.int32).max)
```
Linting is failing because the line is too long

NicolasHug replied:
also this should probably be using `set_params`. we might special-case the way we handle random_state in the future as discussed in https://github.com/scikit-learn/scikit-learn/issues/14042

rth reviewed (commented):
Thanks! It would be good to have a unit test for this behaviour. 

One concern is that this is not fully backward compatible (because the same random state will now produce different output) -- let's wait for more feedback to see if this is fine. I think it might be if we consider this a "bug fix".

jnothman commented:
I've not checked this out yet, but backwards incompatibility is not an
issue as IterativeImputer is experimental

david-cortes commented:
Changed to `set_params` now, and broke the long line to confirm to PEP8 limit.

adrinjalali commented:
shouldn't iterative imputer pass the given random state to the underlying estimator in general?

We shouldn't try to convert it to an integer, we just pass it along (the original `random_state`). If the underlying model doesn't support an rng object, the user shouldn't pass that to IterativeImputer.

I think I'm missing something here though.

NicolasHug commented:
>  If the underlying model doesn't support an rng object, the user shouldn't pass that to IterativeImputer.

That means that by default (i.e. with random_state=None), the IterativeImputer isn't compatible with third party estimators that only support ints.

Is it really part of our convention that estimators must support RandomState and None?

david-cortes commented:
> shouldn't iterative imputer pass the given random state to the underlying estimator in general?
> 
> We shouldn't try to convert it to an integer, we just pass it along (the original `random_state`). If the underlying model doesn't support an rng object, the user shouldn't pass that to IterativeImputer.
> 
> I think I'm missing something here though.

That's precisely the point of the PR: Estimators from outside scikit-learn which are scikit-learn-compatible oftentimes don't work with NumPy's `RandomState` object, which reduces the array of choices a lot, for no good reason. IterativeImputer still makes random choices on its own, so it needs to have a `RandomState` object created inside, even if you pass it an integer for `random_state`.

adrinjalali commented:
If the user passes `None` as the random state to `IterativeImputer`, we don't have to set the random state of the underlying estimator at all. It will stay whatever the default of that estimator is. If that estimator doesn't support `None` then its default should be an int, and that will be the one used.

david-cortes commented:
Updated. Now it will only set the `random_state` of the Estimator to an integer (not to a `RandomState`) if:
* The estimator has attribute `random_state`.
* The estimator's `random_state` is `None`.
* `IterativeImputer` was passed a `random_state` other than `None`.

david-cortes committed:
pass integer random_state to estimator in imputer (#15611)

NicolasHug commented:
> If the user passes None as the random state to IterativeImputer, we don't have to set the random state of the underlying estimator at all.

That's not the semantic we've been using so far. None doesn't mean "no random state", it mean "use the numpy singleton".

adrinjalali commented:
could you please avoid force pushing to the branch? makes it hard to keep track of the changes

adrinjalali commented:
This also needs tests

david-cortes committed:
Added tests for setting estimator random_state inside IterativeImputer

david-cortes commented:
Added tests for the cases in which it should and shouldn't set a seed in the estimator, verifying that the seed is an integer.

adrinjalali commented on sklearn/impute/_iterative.py:
```diff
@@ -558,8 +558,14 @@ def fit_transform(self, X, y=None):
+        ):
+            random_state = self.random_state_.randint(
+                np.iinfo(np.int32).max)
+            self._estimator.set_params(random_state=random_state)
```
This should simply be:

``` python
             self._estimator.set_params(random_state=self.random_state)
```

If the given estimator doesn't support int or rng or None, then user shouldn't pass that as the random state.

david-cortes committed:
don't set estimator random_state to integer when passing RandomState to imputer

david-cortes commented:
Modified to pass to the estimator the same `random_state` that was passed to the imputer instead of drawing a random integer from the internal `RandomState`.

adrinjalali commented on sklearn/impute/tests/test_impute.py:
```diff
@@ -1001,6 +1002,39 @@ def test_iterative_imputer_skip_non_missing(skip_complete):
+
+        def predict(self, X):
+            return np.zeros(X.shape[0])
+    X = np.ones((10, 3), dtype=float)
```
a new line just for clarity

adrinjalali reviewed (approved):
this also needs a whats_new entry, otherwise looks good to me.

david-cortes committed:
Mention changes to random_state in IterativeImputer in whats_new, add whitespace to its test

david-cortes commented:
Added the extra line and comment on the changes in What's New.

david-cortes committed:
Remove spaces from empty line for linter checks

jnothman commented on sklearn/impute/tests/test_impute.py:
```diff
@@ -1001,6 +1002,40 @@ def test_iterative_imputer_skip_non_missing(skip_complete):
+
+        def fit(self, X, y):
+            if self.rs_expect == "RandomState":
+                assert isinstance(self.random_state, np.random.RandomState)
```
can't you check the identity of the RandomState?

david-cortes replied:
It's not possible unfortunately as they contain information about memory address. So something like this would return `False`:
```python
import numpy as np
np.random.RandomState(100) == np.random.RandomState(100)
>>> False
```

thomasjpfan replied:
We can try to check for the state explicitly:

```py
import numpy as np
from numpy.testing import assert_array_equal
rng1 = np.random.RandomState(100)
rng2 = np.random.RandomState(100)

rng1_state = rng1.get_state()
rng2_state = rng2.get_state()

assert rng1_state[0] == rng2_state[0]
assert_array_equal(rng1_state[1], rng2_state[1])
assert_array_equal(rng1_state[2:], rng2_state[2:])
```

jnothman commented on sklearn/impute/tests/test_impute.py:
```diff
@@ -1001,6 +1002,40 @@ def test_iterative_imputer_skip_non_missing(skip_complete):
+     (None, "string", "string"), (408, "string", "string"),
+     (np.random.RandomState(seed=409), "string", "string")]
+)
+def test_iterative_imputer_set_estimator_random_state(
```
I am finding it hard to read this test. I think naming and semantics of `rs_expect` could be improved at least.

adrinjalali commented on sklearn/impute/_iterative.py:
```diff
@@ -558,8 +558,12 @@ def fit_transform(self, X, y=None):
+            and self._estimator.random_state is None
+            and self.random_state is not None
+        ):
+            self._estimator.set_params(random_state=self.random_state)
```
It was like that before, and I asked to change it. I don't think we should get a seed and pass around. If the estimator doesn't support rng, the user shouldn't pass an rng. I don't see why
we should workaround the API limitations of third party libraries.

thomasjpfan replied:
I am happy with the current approach, where the int should be used if `self._estimator` only supports random_state ints.

thomasjpfan replied:
Sorry for bringing this point late. I am wondering if `IterativeImputer` should be setting the `random_state` in `self._estimator` at all. In the example the user can do:

```py
imputer = IterativeImputer(LGBMRegressor(random_state=42), random_state = 456)
```

In this version, `IterativeImputer` would not set `self._estimator.random_state`.

What do you think @david-cortes ?

david-cortes replied:
I guess it's also a very valid approach (not setting the estimator's `random_state`), although it might cause some confusion when one has a seed/state set while the other doesn't. However in that snippet (setting the `random_state` in the estimator before passing it to the imputer), with the changes in this PR, it wouldn't set the estimator's `random_state` either, because it isn't `None`.

thomasjpfan replied:
When looking at metaestimators such as `RegressorChain` of `ClassifierChain`, it has a `random_state` parameter, but does not set the `random_state` of the `base_estimator`. I think not setting `random_state` would be more consistent with the rest of the library.

david-cortes commented on doc/whats_new/v0.23.rst:
```diff
@@ -80,7 +80,16 @@ Changelog
+  `RandomState` object with the estimator, but instead will set its attribute
+  `random_state` if it was passed to the imputer and not to the estimator,
+  allowing to use estimators which do not support `RandomState` objects.
+  :pr: `15636` by :user:`David Cortes` <david-cortes>.
```
Good catch! I see some other entries in there are also not entered this way and don’t render correctly (PR’s 15503 and 15918) - may I attempt to correct them too?

jnothman replied:
sure in a new PR, thanks

david-cortes committed:
shorten description in whats_new, rename parameters in tests

david-cortes commented:
Renamed the parameters in the test and added more line breaks for readability now. Also shortened the description in what's new.

NicolasHug commented on sklearn/impute/tests/test_impute.py:
```diff
@@ -1001,6 +1002,43 @@ def test_iterative_imputer_skip_non_missing(skip_complete):
+)
+def test_iterative_imputer_set_estimator_random_state(
+    rs_imputer, rs_estimator, expect_in_fit
+):
```
Please add a short comment of what this test does

NicolasHug commented on doc/whats_new/v0.23.rst:
```diff
@@ -80,7 +80,14 @@ Changelog
+...............................
+
+- |Enhancement| :class:`impute.IterativeImputer` Will set the estimator's
+  `random_state` to the same `random_state` that was passed to the imputer.
```
in contrast to what? what does this enables?

NicolasHug reviewed (commented):
Do we still need `self.random_state_` now?

david-cortes commented:
> Do we still need `self.random_state_` now?

Yes, it's still needed because the imputer itself also makes random choices such as the column order, so it needs an internal `RandomState` from which to generate random numbers.

david-cortes committed:
better explanation of the random_state logic for imputer

david-cortes commented:
Added a better description to the test, the what's new, and also to the documentation just in case. The merge conflict now is because of this other PR: https://github.com/scikit-learn/scikit-learn/pull/16091 which fixed spacing in the previous entry in what's new.

Not sure how to fix it in this pull request (other than force-pushing), as it doesn't show as a conflict in my local environment.

cmarmo commented:
Hello @david-cortes , if you are still interested in working on this, do you mind fixing conflicts? This will bring more attention from the reviewer side. Thanks for your work!

david-cortes commented:
Solved merge conflicts now.

thomasjpfan commented on doc/whats_new/v0.23.rst:
```diff
@@ -389,6 +389,14 @@ Changelog
   :class:`tree.DecisionTreeRegressor`. :pr:`15864` by
   `Nicolas Hug`_.
 
+:mod:`sklearn.impute`
+...............................
+
+- |Enhancement| :class:`impute.IterativeImputer` Will set the estimator's
+  `random_state` to the same `random_state` that was passed to the imputer
+  instead of a `RandomState`, allowing to use it with more external classes.
+  :pr:`15636` by :user:`David Cortes <david-cortes>`.
```
This needs to move to 0.24 now.

david-cortes replied:
Not sure what you mean by current approach with 'int'. The current approach (as it is in the master branch right now) is to assign a `RandomState` object to the estimator’s `random_state` if it has such attribute.

thomasjpfan replied:
Third party estimators often restrict `random_state` to only be integers. In the current state of this PR, we are passing the integer to `self._estimators`, which I am okay with.

thomasjpfan replied:
If a user were to use a `RandomState` object with an `self._estimator` that does not support `RandomState` objects, this will fail.

david-cortes replied:
Moved the whatsnew notice to 0.24 and turned the test into an equality test of the `RandomState` states.

thomasjpfan reviewed (commented):
Thank you for the PR @david-cortes !

david-cortes committed:
move description about PR to new version

david-cortes committed:
compare states of RandomState objects

cmarmo commented:
@david-cortes , thanks for the updates. The failing check is not related with your PR.
Do you mind fixing conflicts? Then perhaps @thomasjpfan will have some time to check if comments have been addresses? Thanks!

david-cortes commented:
Solved merge conflicts.

thomasjpfan commented on doc/whats_new/v0.24.rst:
```diff
@@ -228,6 +228,14 @@ Changelog
+:mod:`sklearn.impute`
+...............................
+
+- |Enhancement| :class:`impute.IterativeImputer` Will set the estimator's
```
```suggestion
- |Enhancement| :class:`impute.IterativeImputer` will set the estimator's
```

thomasjpfan commented on sklearn/impute/_iterative.py:
```diff
@@ -130,7 +130,9 @@ class IterativeImputer(_BaseImputer):
         selection of estimator features if n_nearest_features is not None, the
         ``imputation_order`` if ``random``, and the sampling from posterior if
-        ``sample_posterior`` is True. Use an integer for determinism.
+        ``sample_posterior`` is True. If the estimator has a `random_state`,
+        but it's set to None, will set it to this value too. Use an integer
```
```suggestion
        ``sample_posterior`` is True. If the estimator has a `random_state`,
        and it is set to None, then it's `random_state` will set it to this value
        too. Use an integer for determinism.
```

thomasjpfan commented on sklearn/impute/tests/test_impute.py:
```diff
@@ -1068,6 +1069,50 @@ def test_iterative_imputer_skip_non_missing(skip_complete):
 
 
+@pytest.mark.parametrize(
+    "rs_imputer, rs_estimator, expect_in_fit",
```
```suggestion
    "rs_imputer, rs_estimator, rs_expect",
```

And the rest of the test funcion.

david-cortes committed:
changes after review suggestions

david-cortes commented:
Incorporated suggestions.

david-cortes committed:
avoid setting estimator's random state at all

david-cortes commented:
Changed the logic of the PR to avoid setting the estimator's `random_state` at all in order to be more consistent with the other similar classes as suggested by @thomasjpfan . Since now it will not do anything with it I removed the test as it doesn't have anything to test now.

david-cortes committed:
remove unneeded imports after removing test

thomasjpfan reviewed (commented):
Thank you for the updates @david-cortes !

@adrinjalali What do you think of the recent changes here?

adrinjalali reviewed (approved):
Strangely enough it seems this change doesn't require a change in the `random_state`'s description. Seems like a _fix_ to me.

Still feels odd not to have a test for this change. We could test that the `IterativeImputer` does not set the `random_state` of a child estimator, otherwise LGTM.

I haven't read the whole conversation, is @NicolasHug happy with the change?

NicolasHug reviewed (approved):
thanks for your patience @david-cortes

david-cortes committed:
change whats_new entry for iterative imputer to 'fix'

david-cortes committed:
test that iterative imputer doesn't set the estimator's random state

david-cortes committed:
shorten parametrization in pytest for random states

cmarmo commented:
It seems to me that the codecov failure is not related to the PR. @thomasjpfan, two approvals: time to merge? :) Thanks!
